### PR TITLE
DEP: rename hanning windows to hann window while keeping compatibility

### LINKED ISSUE #17303: Feature Request: add hann alias for incorrectly named hanning function, update hanning documentation
Currently NumPy has a function, `hanning`, that is documented as "return[ing] the Hanning window". See here:

https://numpy.org/doc/stable/reference/generated/numpy.hanning.html?highlight=hanning#numpy.hanning

This function is actually a misnomer for the [Hann window](https://en.wikipedia.org/wiki/Hann_function). The window is named after Julius von Hann. Its Wikipedia article even mentions this confusion over its name:

"However, the erroneous 'Hanning' function is also heard of on occasion, derived from the paper in which it was named, where the term 'hanning a signal' was used to mean applying the Hann window to it. The confusion arose from the similar Hamming function, named after Richard Hamming."

The NumPy documentation also mentions the discrepancy:

"The Hanning was named for Julius von Hann, an Austrian meteorologist. It is also known as the Cosine Bell. Some authors prefer that it be called a Hann window, to help avoid confusion with the very similar Hamming window."

I don't know what the basis for this latter claim is. [matlab](https://www.mathworks.com/help/signal/ref/hann.html), [SciPy](https://scipy.github.io/devdocs/generated/scipy.signal.windows.hann.html), [TensorFlow](https://www.tensorflow.org/api_docs/python/tf/signal/hann_window) and [PyTorch](https://pytorch.org/docs/master/generated/torch.hann_window.html?highlight=hann#torch.hann_window) use the name `hann` or `hann_window`.  

I suggest adding an alias to satisfy existing NumPy users and those would prefer the correct name.

### PR DESCRIPTION
**Pull request by atsju**
Close #17303

### CODE REVIEW

**Review by jorenham — APPROVED** [review]
As far as the typing changes are concerned, I'm completely on board.

I'm not an expert in signal processing. But assuming that this is indeed the correct name, then this deprecation route seems like the right approach to me.

Thanks!

File: numpy/lib/_function_base_impl.py [hunk 1]
```diff
@@ -3305,17 +3305,19 @@ def hanning(M):
...
-    The Hanning was named for Julius von Hann, an Austrian meteorologist.
-    It is also known as the Cosine Bell. Some authors prefer that it be
-    called a Hann window, to help avoid confusion with the very similar
-    Hamming window.
+    The Hann window was named for Julius von Hann, an Austrian meteorologist.
+    It is also known as the Hanning window, which is a misnomer.
+    However, the term Hanning function is also conventionally used, derived
+    from the paper in which the term hanning a signal was used to mean applying
+    the Hann window to it. It is distinct from the similarly-named Hamming
+    function, named after Richard Hamming.
```
Comment (seberg) [inline numpy/lib/_function_base_impl.py:3318]: It's just two sentences and not sure what the "limit" is, but please don't just copy from wikipedia (wikipedia does require attribution!).

File: numpy/lib/_function_base_impl.py [hunk 2]
```diff
@@ -3381,6 +3383,20 @@ def hanning(M):
...
+@set_module('numpy')
+def hanning(M):
+    """
+    `hanning` is deprecated in NumPy 2.3.
+
+    Please use `hann` instead.
+    """
+    # Deprecated in NumPy 2.3, 2025-06-04
+    warnings.warn(
+        "`hanning` is deprecated. Use `hann` instead.",
```
Comment (seberg) [inline numpy/lib/_function_base_impl.py]:
```suggestion
        "`hanning` is deprecated. Use `hann` instead. (Deprecated NumPy 2.4)",
```
And yes, it is definitely 2.4, not 2.3 at this point.

File: doc/release/upcoming_changes/29123.hannWindow.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+* np.hanning is deprecated. Use np.hann instead.
```
Comment (seberg) [inline doc/release/upcoming_changes/29123.hannWindow.rst]: Use backticks for code (double backticks without link).

**Review by seberg — COMMENTED** [review]
Thanks, but needs a test as everything does.

It may make sense to go the SciPy route, and just add `hann` first and not document `hanning` much, and start deprecation later (the advantage is that libraries can switch to `hann` directly and don't need conditionals).  Although 14k hits on github isn't _that_ much.  I honestly don't have a strong opinion either way.

(For the most part, I think `scipy.signal.windows` is the right place, but these have been around forever and we should keep some... Would be nice to link out to there, but not this PR.)

File: doc/release/upcoming_changes/29123.hannWindow.rst (hunk 3 above)
Comment (atsju) [inline doc/release/upcoming_changes/29123.hannWindow.rst]:
```suggestion
* `np.hanning` is deprecated. Use `np.hann` instead.
```

File: numpy/lib/_function_base_impl.py (hunk 1 above)
Comment (atsju) [inline numpy/lib/_function_base_impl.py:3318]:
```suggestion
    The Hann window was named for Julius von Hann, an Austrian meteorologist.
    It is often mistakenly called the "Hanning" window. This confusion likely
    arises from its similarity to the Hamming window, named after Richard Hamming.
```

Would this be OK ?

**Comment by atsju** [comment]
> Thanks, but needs a test as everything does.
> 
> It may make sense to go the SciPy route, and just add `hann` first and not document `hanning` much, and start deprecation later (the advantage is that libraries can switch to `hann` directly and don't need conditionals). Although 14k hits on github isn't _that_ much. I honestly don't have a strong opinion either way.
> 
> (For the most part, I think `scipy.signal.windows` is the right place, but these have been around forever and we should keep some... Would be nice to link out to there, but not this PR.)

This PR has already been a little personal challenge and I don't know how you maintainers work in this repo. I'm not sure to understand everything here. Just let me know clearly if there is any more to do on my side. (I will not be available during upcoming 2 weeks though).

**Comment by mattip** [comment]
This seems OK, but let's not deprecate the `hanning` function, there are many uses in a quick github code search.
